### PR TITLE
r/api_gateway_domain_name_access_association - new resource

### LINKED ISSUE #40274: API Gateway custom domain names for private REST APIs
> Amazon API Gateway (APIGW) now gives you the ability to manage your private REST APIs using custom user-friendly private DNS name like private.example.com, simplifying API discovery.

* [Announcement](https://aws.amazon.com/about-aws/whats-new/2024/11/amazon-api-gateway-custom-domain-name-private-rest-apis/)
* [Blog post](https://aws.amazon.com/blogs/compute/implementing-custom-domain-names-for-private-endpoints-with-amazon-api-gateway/)
* [Developer Guide](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-private-custom-domains.html)

### PR DESCRIPTION
**Pull request by DrFaust92**
### Relations

Closes #40274
Relates #40364
Relates #40447

### Output from Acceptance Testing

```console
% make testacc TESTS=TestAccAPIGatewayDomainNameAccessAssociation_ PKG=apigateway

--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_basic (157.58s)

```

### CODE REVIEW

File: internal/service/apigateway/domain_name.go [hunk 1]
```diff
@@ -290,6 +294,7 @@ func resourceDomainNameRead(ctx context.Context, d *schema.ResourceData, meta in
 	d.Set("cloudfront_zone_id", meta.(*conns.AWSClient).CloudFrontDistributionHostedZoneID(ctx))
 	d.Set(names.AttrDomainName, output.DomainName)
 	d.Set("domain_name_id", output.DomainNameId)
+	d.Set("domain_name_arn", output.DomainNameArn)
```
Comment (DrFaust92) [inline internal/service/apigateway/domain_name.go:297]: using `arn` doesnt work in upstream APIs. I dont want to break any existing assumptions but it seems to be incorrect so I added the new param. I dont mind removing this and setting the `arn` to expected value. up to maintainers

Comment (ewbankkit) [reply internal/service/apigateway/domain_name.go:297]: Let's try a stick with `arn`.
According to the documentation (e.g. https://docs.aws.amazon.com/apigateway/latest/api/API_GetDomainName.html#apigw-GetDomainName-response-domainNameArn) `output.DomainNameArn` is only set for private custom domain names.
We could do something like

```go
	if output.DomainNameArn != nil {
		d.Set(names.AttrARN, output.DomainNameArn)
	} else {
		d.Set(names.AttrARN, domainNameARN(ctx, meta.(*conns.AWSClient), d.Id()))
	}
```

Comment (ewbankkit) [reply internal/service/apigateway/domain_name.go:297]: And remove the new `domain_name_arn` attribute.

Comment (ewbankkit) [reply internal/service/apigateway/domain_name.go:297]: I am working on this...

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccAPIGatewayDomainNameAccessAssociation_' PKG=apigateway ACCTEST_PARALLELISM=3
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.3 test ./internal/service/apigateway/... -v -count 1 -parallel 3  -run=TestAccAPIGatewayDomainNameAccessAssociation_ -timeout 360m
2024/12/17 08:45:26 Initializing Terraform AWS Provider...
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_null
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_null
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyMap
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyMap
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_AddOnUpdate
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_AddOnUpdate
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnCreate
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnCreate
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnUpdate_Add
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnUpdate_Add
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnUpdate_Replace
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnUpdate_Replace
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_providerOnly
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_providerOnly
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nonOverlapping
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nonOverlapping
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_overlapping
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_overlapping
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_updateToProviderOnly
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_updateToProviderOnly
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_updateToResourceOnly
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_updateToResourceOnly
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_emptyResourceTag
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_emptyResourceTag
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_emptyProviderOnlyTag
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_emptyProviderOnlyTag
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nullOverlappingResourceTag
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nullOverlappingResourceTag
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nullNonOverlappingResourceTag
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nullNonOverlappingResourceTag
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnCreate
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnCreate
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnUpdate_Add
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnUpdate_Add
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnUpdate_Replace
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnUpdate_Replace
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_IgnoreTags_Overlap_DefaultTag
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_IgnoreTags_Overlap_DefaultTag
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_tags_IgnoreTags_Overlap_ResourceTag
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_tags_IgnoreTags_Overlap_ResourceTag
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_basic
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_basic
=== RUN   TestAccAPIGatewayDomainNameAccessAssociation_disappears
=== PAUSE TestAccAPIGatewayDomainNameAccessAssociation_disappears
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_emptyResourceTag
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnUpdate_Replace
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnUpdate_Replace (129.85s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnUpdate_Replace
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags (227.14s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_disappears
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_emptyResourceTag (230.04s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_basic
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnUpdate_Replace (164.03s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_IgnoreTags_Overlap_ResourceTag
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_basic (128.38s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_IgnoreTags_Overlap_DefaultTag
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_IgnoreTags_Overlap_ResourceTag (151.62s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_AddOnUpdate
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_IgnoreTags_Overlap_DefaultTag (167.95s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnUpdate_Add
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_disappears (302.49s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnCreate
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_AddOnUpdate (171.38s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nullNonOverlappingResourceTag
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnCreate (162.90s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnUpdate_Add
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyTag_OnUpdate_Add (190.23s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnCreate
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nullNonOverlappingResourceTag (130.63s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_overlapping
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_updateToResourceOnly
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnUpdate_Add (134.65s)
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_ComputedTag_OnCreate (130.09s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_updateToProviderOnly
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_overlapping (182.01s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nonOverlapping
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_updateToResourceOnly (146.38s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nullOverlappingResourceTag
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_updateToProviderOnly (146.55s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyMap
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nonOverlapping (145.81s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_emptyProviderOnlyTag
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_nullOverlappingResourceTag (137.35s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_null
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_EmptyMap (180.63s)
=== CONT  TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_providerOnly
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_emptyProviderOnlyTag (130.39s)
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_null (96.95s)
--- PASS: TestAccAPIGatewayDomainNameAccessAssociation_tags_DefaultTags_providerOnly (140.01s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/apigateway	1320.571s
```

**Review by YakDriver — APPROVED** [review]

**Comment by ewbankkit** [comment]
@DrFaust92 Thanks for the contribution :tada: :clap:.

**Pull request merged**